src/app/clusters/tls-certificate-management-server/CertificateTableImpl.cpp

```diff
@@ -237,37 +532,168 @@ void CertificateTableImpl::Finish()
+CHIP_ERROR CertificateTableImpl::GetRootCertificateEntry(FabricIndex fabric_index, TLSCAID id, BufferedRootCert & entry)
 {
-    CertificateId id(certificate_id);
-    return mRootCertificates.GetTableEntry(fabric_index, id, entry.mCert, GetBuffer(entry));
+    CertificateId localId(id);
+    return mRootCertificates.GetTableEntry(fabric_index, localId, entry.mCert, GetBuffer(entry));
 }
 
-CHIP_ERROR CertificateTableImpl::HasRootCertificateEntry(FabricIndex fabric_index, TLSCAID certificate_id)
+CHIP_ERROR CertificateTableImpl::HasRootCertificateEntry(FabricIndex fabric_index, TLSCAID id)
 {
-    CertificateId id(certificate_id);
+    CertificateId localId(id);
     EntryIndex unused;
     return mRootCertificates.FindTableEntry(fabric_index, id, unused);
 }
 
-CHIP_ERROR CertificateTableImpl::GetClientCertificateEntry(FabricIndex fabric_index, TLSCCDID certificate_id,
-                                                           BufferedClientCert & entry)
+CHIP_ERROR CertificateTableImpl::IterateRootCertificates(FabricIndex fabric, BufferedRootCert & store,
+                                                         IterateRootCertFnType iterateFn)
+{
+    return mRootCertificates.IterateEntries(fabric, GetBuffer(store), [&](auto & iter) {
+        TableEntryDataConvertingIterator<CertificateId, RootCertStruct> innerIter(iter);
+        return iterateFn(innerIter);
+    });
+}
+
+CHIP_ERROR CertificateTableImpl::RemoveRootCertificate(FabricIndex fabric, TLSCAID id)
+{
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    GlobalCertificateData globalData(mEndpointId);
+    ReturnErrorOnFailure(globalData.Load(mStorage));
+    return globalData.RemoveRootCertificate(*mStorage, mRootCertificates, fabric, id);
+}
+
+CHIP_ERROR CertificateTableImpl::GetRootCertificateCount(FabricIndex fabric, uint8_t & outCount)
 {
-    CertificateId id(certificate_id);
-    return mClientCertificates.GetTableEntry(fabric_index, id, entry.mCert, GetBuffer(entry));
+    return mRootCertificates.GetFabricEntryCount(fabric, outCount);
 }
 
-CHIP_ERROR CertificateTableImpl::HasClientCertificateEntry(FabricIndex fabric_index, TLSCCDID certificate_id)
+CHIP_ERROR CertificateTableImpl::PrepareClientCertificate(FabricIndex fabric, const ByteSpan & nonce, ClientBuffer & buffer,
+                                                          TLSCCDID & id, MutableByteSpan & csr, MutableByteSpan & nonceSignature)
 {
-    CertificateId id(certificate_id);
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    // TODO(gmarcosb): Use PSAKeyAllocator instead
+
+    Crypto::P256Keypair keyPair;
+
+    ReturnErrorOnFailure(keyPair.Initialize(Crypto::ECPKeyTarget::ECDSA));
+
+    size_t csrLen = csr.size();
+    ReturnErrorOnFailure(keyPair.NewCertificateSigningRequest(csr.data(), csrLen));
+    csr.reduce_size(csrLen);
+
+    // Build nocsr_elements_message per spec
+    std::array<uint8_t, 256> nocsrElementsMessage;
```

Comment:
With CSR being `max 3000` and nonce being `128` (this PR, otherwise getting constrained to `32` with upcoming spec changes), I think the buffer here needs to be slightly larger

Reply:
Good call, I think I was thinking about the signature & not the whole CSR 🤦‍♂️ Fixed